### PR TITLE
DB-10044: launch new olap server during online upgrade

### PR DESCRIPTION
**Pull request by jyuanca**

### CODE REVIEW

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1499/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/1501/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by ipraznik-splice — APPROVED** [review]

File: hbase_sql/src/main/java/com/splicemachine/olap/AsyncOlapNIOLayer.java [hunk 1]
```diff
@@ -95,6 +102,19 @@ private void connect() throws IOException {
             HostAndPort hap = hostProvider.olapServerHost(queue);
             LOG.info("Connecting to " + hap);
 
+            if (RegionServerLifecycleObserver.isHbaseJVM) {
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/olap/AsyncOlapNIOLayer.java:105]: What other option is there here?

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java [hunk 2]
```diff
@@ -146,18 +149,22 @@ public void run() {
...
                 try {
                     ApplicationId appId = findApplication(yarnClient);
-                    if (appId == null) {
+                    if (appId == null || (rollingUpgrade && appState != YarnApplicationState.FINISHED)) {
+                        SpliceLogUtils.info(LOG, "Submitting a new OLAP server for rolling upgrade");
                         appId = submitApplication(yarnClient, memory, memoryOverhead, olapPort, cpuCores, sparkYarnQueue, yarnQueue);
                     }
-
+                    else {
+                        // During rolling upgrade, another submitter launched an olap server
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java:165]: This breaks existing logic, if an HMaster restarts (or fails over) we won't keep pinging the still running OlapServer so it'll eventually exit too.

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java [hunk 3]
```diff
@@ -146,18 +149,22 @@ public void run() {
...
+                yarnQueue = conf.get("splice.spark.yarn.queue", YARN_DEFAULT_QUEUE_NAME);
             }
-
+            YarnApplicationState appState = YarnApplicationState.NEW;
             for (int i = 0; i<maxAttempts; ++i) {
                 try {
                     ApplicationId appId = findApplication(yarnClient);
-                    if (appId == null) {
+                    if (appId == null || (rollingUpgrade && appState != YarnApplicationState.FINISHED)) {
+                        SpliceLogUtils.info(LOG, "Submitting a new OLAP server for rolling upgrade");
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java:161]: This is not only called for rolling upgrade, right? Also for the normal OlapServer execution.

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java [hunk 4]
```diff
@@ -598,29 +607,45 @@ static String quoteForBatchScript(String arg) {
     private String prepareCommands(String exec, String parameters, String sparkYarnQueue) throws IOException {
         StringBuilder result = new StringBuilder();
         result.append(exec);
-        for (Object sysPropertyKey : System.getProperties().keySet()) {
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java:601]: This changes how we handle Spark configuration significantly, we need an upgrade path without breaking all existing installations.

**Review by dgomezferro — CHANGES_REQUESTED** [review]

File: hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java (hunk 4 above)
Comment (jyuanca) [reply hbase_sql/src/main/java/com/splicemachine/olap/OlapServerSubmitter.java:601]: This was based on our discussion - Olap server parameters are set as HMaster Java options. This is sometimes confusing for customer. We decided to move everything to hbase-site.xml

File: hbase_sql/src/main/java/com/splicemachine/hbase/SpliceRSRpcServices.java [hunk 5]
```diff
@@ -12,12 +12,15 @@
  * If not, see <http://www.gnu.org/licenses/>.
  */
 
-package com.splicemachine.si.data.hbase.coprocessor;
```
[the remark this answers is not in the document]
Comment (jyuanca) [reply hbase_sql/src/main/java/com/splicemachine/hbase/SpliceRSRpcServices.java:15]: I moved this due to dependencies.

**Comment by jyuanca** [comment]
closing this to open a new one

**Pull request closed without merging**